src/org/mozilla/javascript/Kit.java

```diff
@@ -144,6 +197,52 @@ public static int xDigitToInt(int c, int accumulator)
         return (accumulator << 4) | c;
     }
 
+    public static void reinitNonClassCache() {
+        notClassMap = Collections.synchronizedMap(new WeakHashMap<ClassLoader, LRUCache>());
+    }
+
+    private static synchronized LRUCache getOrCreateCache(ClassLoader loader) {
```

Comment:
It is good that this is synchronized. However, we also have to synchronize usage of the cache once it's been created.

Also, if you're doing this for efficiency, then "notClassMap" could be a ConcurrentHashMap, and then you see if this is cleaner and uses less locking if you use "putIfAbsent". That leads to other complexity though, and it's probably a toss-up whether it buys you anything.

Reply:
I don't believe the JRE has a concurrent version of a WeakHashMap. Google Guava does, but we probably don't want to pull in that dependency.